File: android/ac-actions/src/main/kotlin/com/microsoft/adaptivecards/actions/ActionDelegate.kt
package com.microsoft.adaptivecards.actions

/**
 * Delegate interface for host app to handle action callbacks
 */
interface ActionDelegate {
    /**
     * Called when a Submit action is triggered
     * @param data Map of input IDs to their values
     * @param actionId Optional ID of the action that was triggered
     */
    fun onSubmit(data: Map<String, Any>, actionId: String? = null)

    /**
     * Called when an OpenUrl action is triggered
     * @param url The URL to open
     * @param actionId Optional ID of the action that was triggered
     */
    fun onOpenUrl(url: String, actionId: String? = null)

    /**
     * Called when an Execute action is triggered
     * @param verb The action verb
     * @param data Map of input IDs to their values plus any action data
     * @param actionId Optional ID of the action that was triggered
     */
    fun onExecute(verb: String, data: Map<String, Any>, actionId: String? = null)

    /**
     * Called when a ShowCard action is triggered
     * @param actionId The ID of the action
     * @param isExpanded Whether the card is now expanded
     */
    fun onShowCard(actionId: String, isExpanded: Boolean)

    /**
     * Called when a ToggleVisibility action is triggered
     * @param targetElementIds List of element IDs whose visibility was toggled
     */
    fun onToggleVisibility(targetElementIds: List<String>)
}

/**
 * Default implementation of ActionDelegate that does nothing
 */
class DefaultActionDelegate : ActionDelegate {
    override fun onSubmit(data: Map<String, Any>, actionId: String?) {
        // Default: no-op
    }

    override fun onOpenUrl(url: String, actionId: String?) {
        // Default: no-op
    }

    override fun onExecute(verb: String, data: Map<String, Any>, actionId: String?) {
        // Default: no-op
    }

    override fun onShowCard(actionId: String, isExpanded: Boolean) {
        // Default: no-op
    }

    override fun onToggleVisibility(targetElementIds: List<String>) {
        // Default: no-op
    }
}

File: ios/Sources/ACActions/ActionDelegate.swift
import Foundation
import ACCore

/// Protocol for handling action events from Adaptive Cards
public protocol ActionDelegate: AnyObject {
    /// Called when an Action.Submit is triggered
    func onSubmit(data: [String: Any], actionId: String?)

    /// Called when an Action.OpenUrl is triggered
    func onOpenUrl(url: URL, actionId: String?)

    /// Called when an Action.Execute is triggered
    func onExecute(verb: String?, data: [String: Any], actionId: String?)

    /// Called when an Action.ShowCard is triggered
    func onShowCard(actionId: String?, isExpanded: Bool)

    /// Called when an Action.ToggleVisibility is triggered
    func onToggleVisibility(targetElementIds: [String])
}

/// Default implementation providing empty handlers
public extension ActionDelegate {
    func onSubmit(data: [String: Any], actionId: String?) {
        print("Submit action triggered with data: \(data)")
    }

    func onOpenUrl(url: URL, actionId: String?) {
        print("OpenUrl action triggered with URL: \(url)")
    }

    func onExecute(verb: String?, data: [String: Any], actionId: String?) {
        print("Execute action triggered with verb: \(verb ?? "nil"), data: \(data)")
    }

    func onShowCard(actionId: String?, isExpanded: Bool) {
        // Default: no-op (card expansion handled internally by CardViewModel)
    }

    func onToggleVisibility(targetElementIds: [String]) {
        // Default: no-op (visibility toggling handled internally by CardViewModel)
    }
}
